File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {

        int start = 0;
        int end = nums.length - 1;

        while (start <= end) {
            int middle = (start + end) / 2;
            int startVal = nums[start];
            int currVal = nums[middle];
            if (currVal == target) {
                return middle;
            }
            if (currVal >= startVal) {
                if (target > currVal || target < startVal)
                    start = middle + 1;
                else
                    end = middle - 1;
            } else {
                if (target < currVal || target > nums[end]) {
                    end = middle - 1;
                } else
                    start = middle + 1;
            }
        }

        return -1;
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        start = 0
        end = len(nums)-1

        while(start<=end):
            middle = (start+end)/2
            startVal = nums[start]
            middleVal = nums[middle]
            endVal = nums[end]

            if (target == startVal):
                return start
            if (target == middleVal):
                return middle
            if (target == endVal):
                return end

            if (middleVal>endVal):
                if (target>startVal and target<middleVal):
                    end = middle - 1
                else:
                    start = middle + 1
            elif (middleVal<startVal):
                if (target>middleVal and target<endVal):
                    start = middle + 1
                else:
                    end = middle - 1
            else:
                if (target>middleVal):
                    start = middle + 1
                else:
                    end = middle - 1
        return -1
            


